scrapyd_k8s/object_storage/libcloud_driver.py

```diff
@@ -19,26 +63,73 @@ def __init__(self, config):
             logger.error('Container name is not set in the configuration.')
             raise ValueError('Container name is not set')
 
-        args_envs = config.joblogs_provider_args(self._storage_provider)
+        args_envs = config.joblogs_storage(self._storage_provider)
         args = {}
-        for arg, env in args_envs.items():
-            env_value = os.getenv(env)
-            if env_value is None:
-                logger.error(f"Environment variable '{env}' for argument '{arg}' is not set.")
-                raise ValueError(f"Environment variable '{env}' for argument '{arg}' is not set.")
-            args[arg] = env_value
+        for arg, value in args_envs.items():
+            value_str = str(value)
+            substituted_value = self._substitute_variables(value_str, arg)
+            logger.debug(f"Substituted value for '{arg}': {substituted_value}")
+            args[arg] = substituted_value
 
         driver_class = get_driver(self._storage_provider)
-        print("CLOUD CREDENTIALS")
-        print(args)
         try:
             self.driver = driver_class(**args)
             logger.info(f"Initialized driver for storage provider '{self._storage_provider}'.")
         except Exception as e:
-            logger.exception(f"Failed to initialize driver for storage provider '{self._storage_provider}'.")
+            logger.exception(f"Failed to initialize driver for storage provider '{self._storage_provider}': {e}")
             raise
 
+    def _substitute_variables(self, value, arg_name):
+        """
+        Replaces placeholders in the configuration value with environment variable values.
+
+        Parameters
+        ----------
+        value : str
+            The configuration value possibly containing placeholders.
+        arg_name : str
+            The name of the argument being processed (for logging purposes).
+
+        Returns
+        -------
+        str
+            The value with placeholders replaced by environment variable values.
+
+        Raises
+        ------
+        ValueError
+            If the required environment variable is not set.
+        """
+        def replace_var(match):
+            env_var = match.group(1)
+            env_value = os.getenv(env_var)
+            if env_value is not None:
+                env_value = env_value.strip().strip('"').strip("'")
+                return env_value
+            else:
+                logger.error(f"Environment variable '{env_var}' is not set for argument '{arg_name}'.")
+                raise ValueError(f"Environment variable '{env_var}' is not set for argument '{arg_name}'.")
+
+        result = self.VARIABLE_PATTERN.sub(replace_var, value)
+        return result
```

Comment:
Fine!
Did you think about escaping? What if we want to include `${}` in the value (e.g. a password), or `${X}`. I would expect this could be done by `\${}` resp. `\${X}` (or `$${}` resp. `$${X}`, let's pick what makes most sense in the Python world).

Reply:
Added a line of code to interpret a common syntax to avoid something and consider it as a part of the password
```
\{}
```